fdbserver/ptxn/MessageSerializer.cpp

```diff
@@ -0,0 +1,348 @@
+	serializer.writeItem(subsequenceSerializedMessageItem);
+}
+
+void SubsequencedMessageSerializer::write(const Subsequence& subsequence, StringRef serializedMessage) {
+	write(SubsequenceSerializedMessageItem{ subsequence, serializedMessage });
+}
+
+const Version& SubsequencedMessageSerializer::getCurrentVersion() const {
+	return header.lastVersion;
+}
+
+const Subsequence& SubsequencedMessageSerializer::getCurrentSubsequence() const {
+	return header.lastSubsequence;
+}
+
+void SubsequencedMessageSerializer::completeVersionWriting() {
+	// If a version exists, yet has 0 items (an empty commit), a NOOP is injected to avoid version skipping. Since the
+	// NOOP basically does nothing, it has the maximum subsequence number, which *hopefully* will never be reached.
+	if (serializer.getNumItemsCurrentSection() == 0) {
+		write(MAX_SUBSEQUENCE, MutationRef{ MutationRef::NoOp, ""_sr, ""_sr });
+	}
+
+	serializer.completeSectionWriting();
+
+	details::SubsequencedItemsHeader sectionHeader;
+
+	sectionHeader.version = header.lastVersion;
+	sectionHeader.numItems = serializer.getNumItemsCurrentSection();
+	sectionHeader.length = serializer.getTotalBytes() - header.length;
+
+	serializer.writeSectionHeader(sectionHeader);
+}
+
+void SubsequencedMessageSerializer::completeMessageWriting() {
+	serializer.completeAllItemsWriting();
+
+	header.numItems = serializer.getNumSections();
+	header.length = serializer.getTotalBytes() - serializer.getMainHeaderBytes();
+
+	serializer.writeHeader(header);
+}
+
+Standalone<StringRef> SubsequencedMessageSerializer::getSerialized() {
+	ASSERT(serializer.isAllItemsCompleted());
+
+	return serializer.getSerialized();
+}
+
+size_t SubsequencedMessageSerializer::getTotalBytes() const {
+	return serializer.getTotalBytes();
+}
+
+ProxySubsequencedMessageSerializer::ProxySubsequencedMessageSerializer(const Version& version_) : version(version_) {}
+
+const Version& ProxySubsequencedMessageSerializer::getVersion() const {
+	return version;
+}
+
+void ProxySubsequencedMessageSerializer::broadcastSpanContext(const SpanContextMessage& spanContext) {
+	spanContextMessage = spanContext;
+	storageTeamInjectedSpanContext.clear();
+}
+
+void ProxySubsequencedMessageSerializer::write(const MutationRef& mutation, const StorageTeamID& storageTeamID) {
+	if (serializers.find(storageTeamID) == serializers.end()) {
+		serializers.emplace(storageTeamID, storageTeamID);
+		serializers.at(storageTeamID).startVersionWriting(version);
+	}
+
+	if (spanContextMessage.present()) {
+		if (storageTeamInjectedSpanContext.find(storageTeamID) == storageTeamInjectedSpanContext.end()) {
+			const SpanContextMessage& message = spanContextMessage.get();
+
+			storageTeamInjectedSpanContext.insert(storageTeamID);
+			serializers.at(storageTeamID).write(SubsequenceSpanContextItem{ subsequence++, message });
+		}
+	}
+
+	serializers.at(storageTeamID).write(SubsequenceMutationItem{ subsequence++, mutation });
```

Comment:
Nit: `serializers.at(storageTeamID)` is called multiple times in this function and could be optimized to only once.